src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala

```diff
@@ -1,79 +1,224 @@
+    // The output file
+    lazy val zipFile = writeCollectionTypeZipFile(workspaceNamespace, workspaceName, entityType, entityQueries, metadata, attributeNames)
+
+    // The output to the user
+    lazy val streamingActorRef = actorRefFactory.actorOf(Props(new StreamingActor(ctx, ContentTypes.`application/octet-stream`, entityType + ".zip")))
+    zipFile map { f =>
+      streamingActorRef ! FirstChunk(HttpData.apply(f.byteArray), 0)
+    }
+
+    Future(Done)
+  }
+
+  private def sendRowsAsChunks(actorRef: ActorRef, query: EntityQuery, querySize: Int, entityType: String, headers: IndexedSeq[String], entities: Seq[Entity]): Unit = {
+    val rows = TSVFormatter.makeEntityRows(entityType, entities, headers)
+    val remaining = querySize - query.page + 1
+    // Send headers as the first chunk of data
+    if (query.page == 1) { actorRef ! FirstChunk(HttpData(headers.mkString("\t") + "\n"), remaining)}
+    // Send entities
+    actorRef ! NextChunk(HttpData(rows.map { _.mkString("\t") }.mkString("\n") + "\n"), remaining - 1)
+  }
+
+  private def writeCollectionTypeZipFile(workspaceNamespace: String, workspaceName: String, entityType: String, entityQueries: Seq[EntityQuery], metadata: EntityTypeMetadata, attributeNames: Option[IndexedSeq[String]]): Future[File] = {
+    val entityWriter: ActorRef = actorRefFactory.actorOf(TSVWriterActor.props(entityType, metadata.attributeNames, attributeNames, entityQueries.size))
+    val membershipWriter: ActorRef = actorRefFactory.actorOf(TSVWriterActor.props(entityType, metadata.attributeNames, attributeNames, entityQueries.size))
+    val foldOperation = entityQueries.foldLeft(Future.successful(0, Seq[File]())) { (accumulator, queryGroup) =>
+      for {
+        acc <- accumulator
+        entityBatch <- getEntityBatchFromQueries(entityQueries, workspaceNamespace, workspaceName, entityType)
+        membershipTSV <- (membershipWriter ? WriteMembershipTSV(acc._1, entityBatch)).mapTo[File]
+        entityTSV <- (entityWriter ? WriteEntityTSV(acc._1, entityBatch)).mapTo[File]
+        zip <- writeFilesToZip(entityType, membershipTSV, entityTSV)
+      } yield (acc._1 + 1, Seq(zip))
+    }
+    foldOperation map { files => files._2.head }
+  }
+
+  private def writeFilesToZip(entityType: String, membershipTSV: File, entityTSV: File): Future[File] = {
+    Future {
+      val zipFile = File.newTemporaryDirectory()
+      val membership = zipFile/s"${entityType}_membership.tsv"
+      val entity = zipFile/s"${entityType}_entity.tsv"
+      membershipTSV.copyTo(membership)
+      entityTSV.copyTo(entity)
+      zipFile.zip()
+    }
+  }
+
+  private def getEntityBatchFromQueries(queryGroup: Seq[EntityQuery], workspaceNamespace: String, workspaceName: String, entityType: String): Future[Seq[Entity]] = {
+    Future.sequence(
```

Comment:
If you're doing a `Future.sequence(something.map(...))` it can usually be written as a `Future.traverse`. E.g.:
```
Future.traverse(queryGroup) { query =>
  getEntities(workspaceNamespace, workspaceName, entityType, query)
} map(_.flatten)
```